src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -223,7 +223,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_am_isend_reply(MPIR_Context_id_t context_i
     ucx_hdr.data_sz = data_sz;
 
     if (dt_contig) {
-        send_buf = MPL_malloc(sizeof(ucx_hdr) + am_hdr_sz, MPL_MEM_BUFFER);
+        MPIR_gpu_malloc_host((void **) &send_buf, sizeof(ucx_hdr) + am_hdr_sz);
```

Comment:
This doesn't need to be a GPU registered buffer, but we don't currently have the logic to support different buffer types in the callback. Since this doesn't impact CPU-only performance, it should be OK for now. @yfguo is reworking this code to use GenQ buffers anyhow.

Reply:
@minsii has reported that allocating registered memory with `MPIR_gpu_malloc_host` significantly slows down the performance vs. direct `MPL_malloc` -- https://github.com/pmodels/mpich/issues/5171#issue-839174760

Besides, I believe the `am`-path in ucx is gpu clean -- https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-gpu-ucx/6/

Reply:
If I recall correctly, all ucx gpu failures are direct paths when the buffer size is big -- only those dtp tests with large counts such as `65530` and `262144`

Reply:
This change didn't impact testing, but is technically required, at least for CUDA:

> __host__​cudaError_t cudaFreeHost ( void* ptr )
> Frees the memory space pointed to by hostPtr, which must have been returned by a previous call to cudaMallocHost() or cudaHostAlloc().

The other change in this PR fixed a leak reported with `--enable-g=all`, which is what I went hunting for in the first place. I guess we could drop this one, but it might come back to bite us later.

Reply:
Oh, I see. I guess the issues is we should use `MPL_free(send_buf);` rather than `MPL_gpu_free_host(send_buf);`.

Reply:
Right. I suppose we can add a flag to the request to inform which free function to use. I'll push a new version shortly.